File: extension/lib/aicu.js
/** 在 aicu.cc 页面上下文拉取发评（携带页面 Cookie / WAF 会话） */

const AicuSync = {
  async fetchMid(apiBase) {
    const resp = await fetch(`${apiBase}/api/ingest/bilibili-mid`);
    if (!resp.ok) throw new Error(`获取 B 站 UID 失败: ${resp.status}（请先启动本机 Web 并同步 B 站 Cookie）`);
    const data = await resp.json();
    if (!data.mid) throw new Error("B 站未登录：扩展弹窗先点「同步 Cookie 到本机」");
    return data.mid;
  },

  async checkEnabled(apiBase) {
    const resp = await fetch(`${apiBase}/api/ingest/aicu-status`);
    if (!resp.ok) return;
    const data = await resp.json();
    if (!data.enabled) {
      throw new Error(
        "AICU 导入未开启：在 ~/.osint/config.yaml 设置 ingest.aicu_enabled: true 后重启 Web 服务"
      );
    }
  },

  waitForTabComplete(tabId, timeoutMs = 45000) {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        chrome.tabs.onUpdated.removeListener(onUpdated);
        reject(new Error("打开 aicu.cc 超时，请手动打开 https://www.aicu.cc/ 后重试"));
      }, timeoutMs);
      function onUpdated(id, info) {
        if (id === tabId && info.status === "complete") {
          clearTimeout(timer);
          chrome.tabs.onUpdated.removeListener(onUpdated);
          resolve();
        }
      }
      chrome.tabs.onUpdated.addListener(onUpdated);
    });
  },

  async findOrOpenAicuTab() {
    const patterns = ["*://www.aicu.cc/*", "*://*.aicu.cc/*"];
    const tabs = await chrome.tabs.query({ url: patterns });
    if (tabs.length) return { tabId: tabs[0].id, opened: false };
    const tab = await chrome.tabs.create({ url: "https://www.aicu.cc/", active: true });
    await this.waitForTabComplete(tab.id);
    return { tabId: tab.id, opened: true };
  },

  async fetchPageInTab(tabId, uid, pn, ps = 100) {
    const results = await chrome.scripting.executeScript({
      target: { tabId },
      world: "MAIN",
      func: async (uidArg, pnArg, psArg) => {
        const url = new URL("https://api.aicu.cc/api/v3/search/getreply");
        url.searchParams.set("uid", String(uidArg));
        url.searchParams.set("pn", String(pnArg));
        url.searchParams.set("ps", String(psArg));
        url.searchParams.set("mode", "0");
        url.searchParams.set("keyword", "");
        const resp = await fetch(url.toString(), {
          credentials: "include",
          headers: {
            Accept: "application/json, text/plain, */*",
            Referer: "https://www.aicu.cc/",
          },
        });
        const text = await resp.text();
        return { ok: resp.ok, status: resp.status, text: text.slice(0, 200000) };
      },
      args: [uid, pn, ps],
    });
    const out = results?.[0]?.result;
    if (!out) throw new Error("无法在 aicu.cc 页面执行请求，请重载扩展后重试");
    if (!out.ok) {
      throw new Error(`AICU HTTP ${out.status}：请在 aicu.cc 页面完成人机验证后再点一次`);
    }
    const snippet = (out.text || "").slice(0, 160);
    if (/Just a moment|safeline|cf-browser-verification|challenge-platform/i.test(snippet)) {
      throw new Error("AICU 人机验证未通过：请在已打开的 aicu.cc 标签页手动查一次 UID，再重试扩展按钮");
    }
    try {
      return JSON.parse(out.text);
    } catch (_) {
      throw new Error("AICU 返回非 JSON：请先在 aicu.cc 页面通过验证并能正常查评论后再试");
    }
  },

  async run() {
    const apiBase = await OSINTConfig.getApiBase();
    await this.checkEnabled(apiBase);
    const mid = await this.fetchMid(apiBase);
    const { tabId, opened } = await this.findOrOpenAicuTab();
    if (opened) {
      await new Promise((r) => setTimeout(r, 2000));
    }

    const pages = [];
    let pn = 1;
    let allCount = 0;
    while (pn <= 100) {
      const body = await this.fetchPageInTab(tabId, mid, pn, 100);
      if (body.code !== 0 && body.code != null) {
        throw new Error(`AICU 业务错误 code=${body.code}`);
      }
      const data = body.data || {};
      const cursor = data.cursor || {};
      allCount = cursor.all_count || allCount;
      pages.push(body);
      if (cursor.is_end || !(data.replies || []).length) break;
      pn += 1;
      await new Promise((r) => setTimeout(r, 1500));
    }

    if (!pages.length || !pages.some((p) => (p.data?.replies || []).length)) {
      throw new Error(`AICU 无发评数据（UID=${mid}）。确认 aicu.cc 能查到该 UID 的评论`);
    }

    const resp = await fetch(`${apiBase}/api/ingest/aicu-json`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ pages }),
    });
    const result = await resp.json().catch(() => ({}));
    if (!resp.ok) {
      throw new Error(result.detail || result.error || resp.statusText);
    }
    if (result.ok === false) {
      if (result.error === "aicu_disabled") {
        throw new Error("服务端 AICU 未开启：config.yaml 设 ingest.aicu_enabled: true 并重启 Web");
      }
      throw new Error(result.error || "导入失败");
    }
    return { mid, pages: pages.length, all_count: allCount, ...result };
  },
};

File: src/osint_toolkit/ingest/aicu.py
"""AICU 第三方 B 站发评历史导入 / Posted-comment history via aicu.cc."""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any

import httpx

from osint_toolkit.http.client import HttpClient
from osint_toolkit.http.ssrf import assert_public_http_url
from osint_toolkit.ingest.bilibili_account import _nav_mid
from osint_toolkit.storage.sqlite import connect
from osint_toolkit.utils.config import load_config

AICU_GETREPLY = "https://api.aicu.cc/api/v3/search/getreply"
_CF_MARKERS = ("Just a moment", "safeline", "cf-browser-verification", "challenge-platform")
logger = logging.getLogger(__name__)


def parent_url_from_dyn(dyn: dict[str, Any] | None, *, aid_bvid: dict[str, str] | None = None) -> str:
    """Map AICU dyn {oid, type} to a content URL."""
    if not dyn:
        return ""
    oid = str(dyn.get("oid") or "")
    if not oid:
        return ""
    ctype = int(dyn.get("type") or 0)
    cache = aid_bvid or {}
    if ctype == 1:
        bvid = cache.get(oid)
        if bvid:
            return f"https://www.bilibili.com/video/{bvid}"
        return f"https://www.bilibili.com/video/av{oid}"
    if ctype == 12:
        return f"https://www.bilibili.com/read/cv{oid}"
    if ctype == 17:
        return f"https://www.bilibili.com/opus/{oid}"
    return ""


def parse_aicu_reply(
    reply: dict[str, Any],
    *,
    aid_bvid: dict[str, str] | None = None,
) -> dict[str, Any] | None:
    rpid = str(reply.get("rpid") or "")
    message = str(reply.get("message") or "").strip()
    if not rpid:
        return None
    dyn = reply.get("dyn") if isinstance(reply.get("dyn"), dict) else {}
    parent_url = parent_url_from_dyn(dyn, aid_bvid=aid_bvid)
    title = message[:200] if message else f"评论 rpid={rpid}"
    return {
        "source": "bilibili",
        "title": title,
        "url": parent_url,
        "message": message,
        "rpid": rpid,
        "parent_url": parent_url,
        "comment_time": int(reply.get("time") or 0),
        "event_kind": "comment_post",
        "via": "aicu",
    }


def parse_aicu_page(body: dict[str, Any]) -> tuple[list[dict[str, Any]], bool, int]:
    """Return (replies, is_end, all_count)."""
    if body.get("code") not in (0, None):
        return [], True, 0
    data = body.get("data") or {}
    cursor = data.get("cursor") or {}
    is_end = bool(cursor.get("is_end"))
    all_count = int(cursor.get("all_count") or 0)
    raw = data.get("replies") or []
    return [r for r in raw if isinstance(r, dict)], is_end, all_count


def extract_replies_from_payload(payload: Any) -> list[dict[str, Any]]:
    """Accept AICU API page JSON, array of pages, or bare replies list."""
    if isinstance(payload, list):
        if payload and isinstance(payload[0], dict) and "rpid" in payload[0]:
            return [r for r in payload if isinstance(r, dict)]
        replies: list[dict[str, Any]] = []
        for item in payload:
            if isinstance(item, dict):
                replies.extend(parse_aicu_page(item)[0])
        return replies
    if isinstance(payload, dict):
        if "replies" in payload and isinstance(payload.get("replies"), list):
            return [r for r in payload["replies"] if isinstance(r, dict)]
        return parse_aicu_page(payload)[0]
    return []


def _dedup_key(event_type: str, rpid: str) -> str:
    return f"{event_type}|{rpid}"


def _try_mark_dedup(conn, dedup_key: str, event_type: str) -> bool:
    cur = conn.execute(
        "INSERT OR IGNORE INTO event_dedup (dedup_key, event_type) VALUES (?, ?)",
        (dedup_key, event_type),
    )
    return cur.rowcount > 0


def _aicu_request_headers() -> dict[str, str]:
    cfg = load_config()
    http_cfg = cfg.get("http", {})
    ingest_cfg = cfg.get("ingest", {})
    ua = str(
        http_cfg.get(
            "user_agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36 Edg/131.0.0.0",
        )
    )
    headers = {
        "User-Agent": ua,
        "Accept": "application/json, text/plain, */*",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        "Referer": "https://www.aicu.cc/",
        "Origin": "https://www.aicu.cc",
    }
    cookie = str(ingest_cfg.get("aicu_cookie") or "").strip()
    if cookie:
        headers["Cookie"] = cookie
    return headers


def _is_waf_block(status_code: int, text: str) -> bool:
    if status_code in {403, 468, 503}:
        return True
    snippet = (text or "")[:500]
    return any(marker in snippet for marker in _CF_MARKERS)


async def _fetch_aicu_page(
    client: httpx.AsyncClient,
    *,
    uid: int,
    pn: int,
    ps: int,
) -> dict[str, Any]:
    headers = _aicu_request_headers()
    assert_public_http_url(AICU_GETREPLY)
    resp = await client.get(
        AICU_GETREPLY,
        params={"uid": str(uid), "pn": str(pn), "ps": str(ps), "mode": "0", "keyword": ""},
        headers=headers,
        timeout=30.0,
    )
    text = resp.text
    if _is_waf_block(resp.status_code, text):
        raise RuntimeError(
            "aicu_waf_blocked: AICU 拦截了程序访问（网页能查、脚本不能）。"
            "请用扩展「浏览器拉取 AICU 发评」或粘贴 Network 里的 JSON 导入。"
        )
    resp.raise_for_status()
    try:
        return resp.json()
    except json.JSONDecodeError as exc:
        if _is_waf_block(resp.status_code, text):
            raise RuntimeError("aicu_waf_blocked: 响应非 JSON，可能被 WAF 拦截") from exc
        raise


async def _resolve_bvid(client: HttpClient, aid: str, cache: dict[str, str]) -> None:
    if aid in cache:
        return
    try:
        resp = await client.get(f"https://api.bilibili.com/x/web-interface/view?aid={aid}")
        payload = resp.json()
        if payload.get("code") == 0:
            bvid = (payload.get("data") or {}).get("bvid") or ""
            if bvid:
                cache[aid] = bvid
    except Exception as exc:  # noqa: BLE001
        logger.warning("aicu: resolve bvid for aid=%s failed: %s", aid, exc)


async def _persist_replies(
    replies: list[dict[str, Any]],
    *,
    limit: int,
    bili: HttpClient | None = None,
) -> tuple[list[dict[str, Any]], int, int]:
    """Return (saved_rows, skipped, all_count_hint).

    先在网络阶段解析所有 aid→bvid 映射并构造 entry，再打开 SQLite 连接执行
    dedup 与写入；避免持有 DB 连接跨 ``await`` 网络请求（AGENTS.md SQLite 并发规约）。
    """
    bili = bili or HttpClient()
    aid_bvid: dict[str, str] = {}
    entries: list[dict[str, Any]] = []
    for raw in replies:
        dyn = raw.get("dyn") if isinstance(raw.get("dyn"), dict) else {}
        if int(dyn.get("type") or 0) == 1:
            oid = str(dyn.get("oid") or "")
            if oid and oid not in aid_bvid:
                await _resolve_bvid(bili, oid, aid_bvid)
        if len(entries) >= limit:
            break
        entry = parse_aicu_reply(raw, aid_bvid=aid_bvid)
        if entry:
            entries.append(entry)
    results: list[dict[str, Any]] = []
    skipped = 0
    conn = connect()
    try:
        for entry in entries:
            dedup_key = _dedup_key("bilibili_comment_post", entry["rpid"])
            if not _try_mark_dedup(conn, dedup_key, "bilibili_comment_post"):
                skipped += 1
                continue
            conn.execute(
                "INSERT INTO events (event_type, data_json) VALUES (?, ?)",
                ("bilibili_comment_post", json.dumps(entry, ensure_ascii=False)),
            )
            results.append(entry)
        conn.commit()
    finally:
        conn.close()
    return results, skipped, len(replies)


async def get_bilibili_mid() -> int | None:
    return await _nav_mid(HttpClient())


async def ingest_aicu_from_json(payload: Any, *, limit: int = 10_000) -> dict[str, Any]:
    from osint_toolkit.utils.config import get_aicu_enabled

    if not get_aicu_enabled():
        return {"ok": False, "error": "aicu_disabled", "count": 0}

    replies = extract_replies_from_payload(payload)
    if not replies:
        return {"ok": False, "error": "aicu_json_empty", "count": 0}

    results, skipped, all_count = await _persist_replies(replies, limit=limit)
    return {
        "ok": True,
        "count": len(results),
        "skipped": skipped,
        "all_count": all_count,
        "source": "json",
        "rows": results[:20],
    }


async def ingest_aicu_comments(
    *,
    limit: int = 10_000,
    page_size: int | None = None,
    delay_sec: float | None = None,
) -> dict[str, Any]:
    from osint_toolkit.utils.config import get_aicu_enabled

    cfg = load_config().get("ingest", {})
    if not get_aicu_enabled():
        return {"ok": False, "error": "aicu_disabled", "count": 0}

    ps = int(page_size or cfg.get("aicu_page_size") or 100)
    delay = float(delay_sec if delay_sec is not None else cfg.get("aicu_delay_sec", 1.5))

    bili = HttpClient()
    mid = await _nav_mid(bili)
    if not mid:
        return {"ok": False, "error": "bilibili_not_logged_in", "count": 0}

    all_replies: list[dict[str, Any]] = []
    pn = 1
    all_count = 0

    try:
        async with httpx.AsyncClient(follow_redirects=True) as client:
            while len(all_replies) < limit:
                body = await _fetch_aicu_page(client, uid=mid, pn=pn, ps=ps)
                replies, is_end, all_count = parse_aicu_page(body)
                if not replies and pn == 1:
                    break
                all_replies.extend(replies)
                if is_end or not replies:
                    break
                pn += 1
                if delay > 0:
                    await asyncio.sleep(delay)
    except Exception as exc:  # noqa: BLE001
        msg = str(exc)
        if "aicu_waf_blocked" in msg:
            return {
                "ok": False,
                "error": "aicu_waf_blocked",
                "detail": msg,
                "count": 0,
                "uid": mid,
                "hint": "先在浏览器打开 aicu.cc 查一次 UID，再用扩展拉取或粘贴 JSON",
            }
        return {"ok": False, "error": f"aicu_request_failed: {exc}", "count": 0, "uid": mid}

    results, skipped, _ = await _persist_replies(all_replies[:limit], limit=limit, bili=bili)
    return {
        "ok": True,
        "count": len(results),
        "skipped": skipped,
        "all_count": all_count,
        "uid": mid,
        "rows": results[:20],
    }


async def probe_aicu() -> dict[str, Any]:
    """探测 AICU API 是否可用（与 scripts/probe_aicu.py 逻辑一致）。"""
    from osint_toolkit.utils.config import get_aicu_enabled

    if not get_aicu_enabled():
        return {"status": "DISABLE", "reason": "sync.aicu_enabled / ingest.aicu_enabled 未开启"}

    mid = await get_bilibili_mid()
    if not mid:
        return {"status": "DISABLE", "reason": "B站未登录，无法探测 AICU"}

    headers = _aicu_request_headers()
    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            resp = await client.get(
                AICU_GETREPLY,
                params={"uid": str(mid), "pn": "1", "ps": "5", "mode": "0", "keyword": ""},
                headers=headers,
            )
            text = resp.text[:500]
            if _is_waf_block(resp.status_code, text):
                return {
                    "status": "WAF_BLOCKED",
                    "reason": "AICU 被 WAF 拦截，请用扩展或粘贴 JSON",
                    "http_status": resp.status_code,
                    "mid": mid,
                }
            data = resp.json()
            replies = (data.get("data") or {}).get("replies") or []
            if data.get("code") not in (0, None):
                return {
                    "status": "FAIL",
                    "reason": data.get("message", "unknown"),
                    "code": data.get("code"),
                    "mid": mid,
                }
            return {"status": "PASS", "mid": mid, "sample_count": len(replies)}
        except json.JSONDecodeError:
            return {"status": "WAF_BLOCKED", "reason": "响应非 JSON，可能被 WAF 拦截", "mid": mid}
        except Exception as exc:  # noqa: BLE001
            return {"status": "FAIL", "reason": str(exc), "mid": mid}
